[MRG + 2] Add Drop Option to OneHotEncoder.

drewmjohnston opened the pull request:
#### Reference Issues/PRs

Fixes #6488 and fixes #6053. This builds upon some of the code from #12884 (thanks @NicolasHug!), but also incorporates functionality which lets the user manually specify which category in each column they would like to be dropped, so this is a more general solution along the lines of what @amueller suggested in #6053. This is useful in some cases (such as OLS regression) where the dropped group affects the interpretation of coefficients. 
This should also fix #9361, which has less functionality and appears stalled.

#### What does this implement/fix? Explain your changes.
This code implements a new parameter (drop) in the OneHotEncoder, which can take any of three values: 
None (default), which implements the existing behavior. 
'first', which drops the first value in each category.
list of length n_features, which allows for the manual specification of the reference group. 

#### Any other comments?

This new feature does not work in Legacy mode (this was discussed in #6053), and it requires the manual specification of "categories='auto'" in the case in which the input is all integers, so as not to interfere with the ongoing change in the treatment of integers in OneHotEncoder.
This code is also incompatible with "handle_missing='ignore'", since in that case it is not possible to determine which categories are 0 because they are in the reference category and which are all 0 due to missing data.

Closes #12884

Commit:
Added documentation for drop one OneHotEncoder

Commit:
Added functionality for drop first and drop manually-specified.

Commit:
Added tests for OneHotEncoder drop parameter

Commit:
Fixed flake8

Commit:
Resolved merge conflict in preprocessing/_encoders.py

Commit:
Added additional code to detect when a column that ought to be dropped is not present in the training data. Added tests to confirm

Commit:
Fixed docstring

Commit:
Fixed docstring2

Commit:
Finally will clear pytest

Commit:
Removed code that is not compatible with numpy 1.11.0

Commit:
Fixed docs to match current OneHotEncoder string

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -161,6 +161,19 @@ class OneHotEncoder(_BaseEncoder):
+        column of ones.
+
+        - None : retain all features.
+        - 'first' : drop the first feature in each category.
```
feature in each category -> category in each feature

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -161,6 +161,19 @@ class OneHotEncoder(_BaseEncoder):
+
+        - None : retain all features.
+        - 'first' : drop the first feature in each category.
+        - array : ``drop[i]`` is the value of the feature in ``X[:,i]``
```
value of the feature -> category to drop

jnothman replied:
Should it be possible for the user to specify that some features do not drop? Perhaps that's unnecessary complexity, and the current design is good.

drewmjohnston replied:
I added this functionality as well, it was simple to implement.

drewmjohnston replied:
sure--changed to category.

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -161,6 +161,19 @@ class OneHotEncoder(_BaseEncoder):
+        features cause problems, such as when feeding the resulting data
+        into a neural network or an unregularized regression. If only one
+        feature appears in a column, that column will be retained as a
+        column of ones.
```
Why should it be retained?

drewmjohnston replied:
No real preference here. I've changed it so it will be dropped instead.

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -210,6 +223,10 @@ class OneHotEncoder(_BaseEncoder):
         of ``transform``).
 
+    drop_ : array
+        The category for each feature to be dropped. None if all features
```
reorder: "to be dropped for each feature"

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -260,6 +277,17 @@ class OneHotEncoder(_BaseEncoder):
            [None, 2]], dtype=object)
     >>> enc.get_feature_names()
     array(['x0_Female', 'x0_Male', 'x1_1', 'x1_2', 'x1_3'], dtype=object)
+    >>> drop_enc = OneHotEncoder(drop='first')
```
Don't need to show the repr again, so do .fit on this line

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -210,6 +223,10 @@ class OneHotEncoder(_BaseEncoder):
         (in order of the features in X and corresponding with the output
         of ``transform``).
```
Note that drop categories are included here (although this does make it hard to identify the correspondence between columns and categories so I'm not entirely sure that's the right design)

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -354,21 +383,39 @@ def _handle_deprecations(self, X):
-                    warnings.warn(msg, FutureWarning)
-                    self._legacy_mode = True
-                    self._n_values = 'auto'
+                    if self.drop is None:
```
I'm okay with, rather, `drop` use also triggering the new mode. Why not?

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -426,6 +488,42 @@ def fit(self, X, y=None):
+                missing_drops = [(val, i) for i, val in enumerate(self.drop)
+                                 if val not in self.categories_[i]]
+                if any(missing_drops):
+                    msg = ("The following features were supposed to be "
```
features -> categories

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -426,6 +488,42 @@ def fit(self, X, y=None):
+                           "of features ({}), got {}")
+                    raise ValueError(msg.format(len(self.categories_),
+                                                len(self.drop)))
+                missing_drops = [(val, i) for i, val in enumerate(self.drop)
```
I think it's confusing that you reverse the tuple

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -426,6 +488,42 @@ def fit(self, X, y=None):
+                    msg = ("The following features were supposed to be "
+                           "dropped, but were not found in the training "
+                           "data.\n{}".format(
+                               "\n".join(["Val: {}, Col: {}".format(v, c)
```
Val -> category
Col -> Feature

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -210,6 +223,10 @@ class OneHotEncoder(_BaseEncoder):
         (in order of the features in X and corresponding with the output
         of ``transform``).
 
+    drop_ : array
```
shape?

jnothman replied:
I think the implementation can be simplified with little harm to the API, if we make it `drop_idx_` such that `categories_[i][drop_idx_[i]]` will be dropped for all `i`. Then fitting `drop='first'` just involves `drop_idx_ = np.zeros(len(self.categories_))`. At transform time the masked column indices are `drop_idx_ + np.cumsum([0] + [len(c) for c in categories_[:-1]])`.

drewmjohnston replied:
This is a good idea, I think I'll update the implementation to reflect this.

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -426,6 +488,42 @@ def fit(self, X, y=None):
+                                         for v, c in missing_drops])))
+                    raise ValueError(msg)
+                self.drop_ = np.array([cat for i, cat in enumerate(self.drop)
+                                       if len(self.categories_[i]) > 1],
```
It's a bit weird to just have entries missing (and hence unaligned to `categories_`) if there's only one category in that column. Especially so if the user has specified `drop` rather than 'first' as they have here. Looking at transform below, I find it hard to believe that this behaviour is properly tested.

drewmjohnston replied:
OK. I've resolved this by incorporating your suggestion from elsewhere in the thread. Now, columns with only a single category are dropped entirely from the transformed version (unless it is manually specified that no category should be dropped from the feature).

jnothman reviewed (commented):
Not yet looked in detail at `transform` or tests. Nice work!

drewmjohnston commented:
Thanks for the feedback @jnothman! I just got back from a vacation and am working my way through these.

Commit:
Updated error message with language that addresses new drop functionality

Commit:
Updated documentation and testing to resolve errors mentioned by jnothman

qinhanmin2014 commented:
When do we need to specify which features to drop? Maybe #12884 is enough?

drewmjohnston commented:
@qinhanmin2014 The main appeal of specifying what feature to drop is for situations in which the coefficient interpretation is important (and done relative to the dropped category), for example in unregularized regression.

Commit:
Fixed Flake8 bug

amueller commented:
I think it would be good to support what feature to drop. If you do this in regularized linear regression, picking which to drop will actually change the optimization problem.

qinhanmin2014 commented:
> @qinhanmin2014 The main appeal of specifying what feature to drop is for situations in which the coefficient interpretation is important (and done relative to the dropped category), for example in unregularized regression.

> I think it would be good to support what feature to drop. If you do this in regularized linear regression, picking which to drop will actually change the optimization problem.

Thanks, now I agree that this one is better than #12884

Commit:
Added new way to encode features to be dropped. Added support for leaving all values of some features. Updated behavior to by default drop columns with same value throughout

Commit:
Merge branch 'master' into leave_one_out_ohe

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -208,6 +224,10 @@ class OneHotEncoder(_BaseEncoder):
         of ``transform``).
 
+    drop_idx_ : array of shape (n_features,)
+        The index in ``categories_ of`` the category to be dropped for
```
"of" should be outside the markup

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -652,17 +779,31 @@ def inverse_transform(self, X):
+            if self.drop is None:
+                cats = self.categories_[i]
+            else:
+                cats = np.array([cat for idx, cat in
```
Can't you just do `np.delete(self.categories_[i], self.drop_idx_[i])`?

Do you need to handle the drop_idx_[i] is None case here?

drewmjohnston replied:
Sure, that makes more sense.

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -636,7 +754,16 @@ def inverse_transform(self, X):
+            n_transformed_features = sum(len(cats)
+                                         for cats in self.categories_)
+        else:
+            n_transformed_features = sum(len(cats) - 1
```
can just do `sum(len(cats) - (drop_idx is not None) for cats, drop_idx in zip(self.categories_, self.drop_idx_)`

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -572,11 +671,30 @@ def _transform_new(self, X):
+            n_values = [len(cats) - 1
+                        if self.drop_idx_[i] is not None
+                        else len(cats)
+                        for i, cats in enumerate(self.categories_)]
```
use zip rather than enumerate?

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -572,11 +671,30 @@ def _transform_new(self, X):
+                drop_value = self.drop_idx_[i]
+                """
+                Add the cells where the drop value is present to the list
+                of those to be masked-out. Decrement the indices of the values
```
I see three options for implementing `transform`:
1. ignore the added category when encoding, a bit like handle_unknown='ignore'
2. remove the value once encoded as an int
3. remove the columns once one-hot encoded

I think you've implemented #2. Why is this one chosen? What are its pros and cons in terms of code simplicity and computational complexity?

drewmjohnston replied:
To do the first option, it seems like I'd have to modify _transform(), which is in the BaseEncoder class. We're not adding drop options to the other classes that inherit from BaseEncoder, so I thought that changing this in BaseEncoder might not be the best approach. 
The approach I took with 2) had some advantages, since I was able to mix it in to reuse some of the logic that was already used for masking unknown values. It also avoids having to do column slicing on an output that might be a CSR matrix (as would be the case in 3)).

drewmjohnston replied:
@jnothman do you think that one of the other approaches is superior?

Commit:
Implemented basic text corrections and code optimizations recommended

Commit:
Merge branch 'master' into leave_one_out_ohe

Commit:
fixed Flake8 Literal complaint

Commit:
Merge branch 'master' into leave_one_out_ohe

drewmjohnston commented:
@jnothman any further comments?

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -572,11 +671,30 @@ def _transform_new(self, X):
+                above so as not to leave a blank column.
+                """
+                if drop_value is not None:
+                    keep_cells = ~np.in1d(Xii, [drop_value])
```
More so, is it possible to do that for all features?
```py
keep_cells = X_int != self.drop_idx_.reshape(1, -1)
X_mask &= keep_cells
X_int[X_int > self.drop_idx_.reshape(1, -1)] -= 1
```

drewmjohnston replied:
Good point with the first bit. I think this code can be simplified at least to:

`for i in range(n_features):
                Xii = X_int[:, i]
                Xmi = X_mask[:, i]
                drop_value = self.drop_idx_[i]
                """
                Add the cells where the drop value is present to the list
                of those to be masked-out. Decrement the indices of the values
                above so as not to leave a blank column.
                """
                if drop_value is not None:
                    Xmi &= np.not_equal(Xii, drop_value)
                    Xii[Xii > drop_value] -= 1
`

Regarding the second bit, I don't think that will handle cases in which one of the features has None for the category to be dropped. Is there a better way to do this than looping over columns? Having to mask the columns makes the code longer and inplace modification tricky.

jnothman replied:
If at this point we are working with integer encoding of X surely it's easy to replace None with an appropriately impossible int?

drewmjohnston replied:
Surely it is, how's this:

`
if self.drop is not None:
    to_drop = np.copy(self.drop_idx_).reshape(1, -1)
    to_drop[to_drop == None] = max(len(cats)
                                   for cats in self.categories_)
    keep_cells = X_int != to_drop
    X_mask &= keep_cells
    X_int[X_int > to_drop] -= 1
    n_values = [len(cats) - 1 if drop_val is not None
                else len(cats)
                for cats, drop_val in
                zip(self.categories_, self.drop_idx_)]
else:
    n_values = [cats.shape[0] for cats in self.categories_]
`

jnothman reviewed (commented):
Sorry this is only a partial review

Commit:
Merge branch 'master' into leave_one_out_ohe

jnothman commented:
Looks almost good except that I'd be careful about comparing to None using ==

Commit:
Removed unnecessary loop on transform step

drewmjohnston commented:
How's:
```
if self.drop is not None:
    max_val = max(len(cats) for cats in self.categories_)
    to_drop = np.array([pos if pos is not None else max_val
                        for pos in self.drop_idx_]).reshape(1, -1)
    keep_cells = X_int != to_drop
    X_mask &= keep_cells
    X_int[X_int > to_drop] -= 1
    n_values = [len(cats) - 1 if drop_val is not None
                else len(cats)
                for cats, drop_val in
                zip(self.categories_, self.drop_idx_)]
else:
    n_values = [cats.shape[0] for cats in self.categories_]
```

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -425,6 +487,44 @@ def fit(self, X, y=None):
             return self
         else:
             self._fit(X, handle_unknown=self.handle_unknown)
+            if self.drop is None:
```
let's pull this out into a method. `self.drop_idx_ = self._compute_drop_idx()`

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -571,11 +671,23 @@ def _transform_new(self, X):
 
+        if self.drop is not None:
+            max_val = max(len(cats) for cats in self.categories_)
+            to_drop = np.array([pos if pos is not None else max_val
```
I prefer not negating if-else operator condition. `max_val if pos is None else pos`

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -571,11 +671,23 @@ def _transform_new(self, X):
+                                for pos in self.drop_idx_]).reshape(1, -1)
+            keep_cells = X_int != to_drop
+            X_mask &= keep_cells
+            X_int[X_int > to_drop] -= 1
```
put a comment here

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -207,6 +224,10 @@ class OneHotEncoder(_BaseEncoder):
         (in order of the features in X and corresponding with the output
         of ``transform``).
```
Note that `categories_` includes the dropped category.

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -635,7 +747,13 @@ def inverse_transform(self, X):
         n_features = len(self.categories_)
-        n_transformed_features = sum([len(cats) for cats in self.categories_])
+        if self.drop is None:
+            n_transformed_features = sum(len(cats)
```
this is repeated from above with `n_values`. Refactor it.

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -651,17 +769,29 @@ def inverse_transform(self, X):
-                unknown = np.asarray(sub.sum(axis=1) == 0).flatten()
-                if unknown.any():
+            X_tr[:, i] = cats[labels]
+            unknown = np.asarray(sub.sum(axis=1) == 0).flatten()
```
I think you want `~sub.any(axis=1)`

drewmjohnston replied:
Hrm, this doesn't play nicely with CSR matrices. Do you think it's worth converting a sparse representation to a dense one to use this?

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -651,17 +769,29 @@ def inverse_transform(self, X):
 
         for i in range(n_features):
-            n_categories = len(self.categories_[i])
+            if (self.drop is None) or self.drop_idx_[i] is None:
```
rm unnecessary parentheses

jnothman commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -373,21 +387,25 @@ def test_one_hot_encoder(X):
 
+    if drop is None:
         # with unknown categories
+        # drop_first is incompatible with handle_unknown=ignore
```
`drop_first` -> `drop`

jnothman commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -687,3 +705,63 @@ def test_one_hot_encoder_warning():
     X = [['Male', 1], ['Female', 3]]
     np.testing.assert_no_warnings(enc.fit_transform, X)
+
+    enc = OneHotEncoder(drop=['def', 3, 56])
```
what's this got to do with testing the warning?

drewmjohnston replied:
I think I accidentally deleted the header line for the next test!

jnothman reviewed (commented):
The tests should include tests of `drop_idx_` type, dtype and value. They should also check that `categories_` is unaffected

Commit:
Added additional tests for categories. Refactored some code in transform steps

Commit:
Merge branch 'master' into leave_one_out_ohe

drewmjohnston commented:
> The tests should include tests of `drop_idx_` type, dtype and value. They should also check that `categories_` is unaffected

What tests exactly do you want for `drop_idx_` type and dtype? It should always be a numpy array (there's only two ways for it to get set), but I think the dtype will just be object in many cases (due to Nones for some categories).

jnothman commented:
> What tests exactly do you want for `drop_idx_` type and dtype? It should always be a numpy array (there's only two ways for it to get set), but I think the dtype will just be object in many cases (due to Nones for some categories).

Then it should probably always be object. Or it can be a list rather than array?

Commit:
Added dtype, type checking for drop_idx_

drewmjohnston commented:
OK, just made it so it enforces dtype=object and checks for it (and type is np.ndarray) in the tests.

jnothman reviewed (approved):
Nice. Thanks!

drewmjohnston commented:
@jnothman great--any idea what's going on with Appveyor? It's throwing an error:
```
Error uploading artifact the storage: The remote name could not be resolved: 'appveyorcidata.blob.core.windows.net'
```

jnothman commented:
Random network failure. Should go green with a restart.

Commit:
Fixed small typo. Repush due to random network error

drewmjohnston commented:
That seems to have worked!

jorisvandenbossche commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -651,17 +776,29 @@ def inverse_transform(self, X):
+            if unknown.any():
+                '''
+                In this case, we can safely assume that all of the nulls in
+                each column are the dropped value
```
Can you use standard `#` comment here?

jorisvandenbossche commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -651,17 +776,29 @@ def inverse_transform(self, X):
             labels = np.asarray(_argmax(sub, axis=1)).flatten()
-            X_tr[:, i] = self.categories_[i][labels]
-
-            if self.handle_unknown == 'ignore':
```
instead of removing this check, you could also add a `or if self.drop is not None` ? (although I don't know how expensive the calculation of unknown is, so how important it is to avoid)

jorisvandenbossche commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -687,3 +705,95 @@ def test_one_hot_encoder_warning():
+    exp = np.array([[1., 0., 1., 1.],
+                    [0., 1., 0., 1.],
+                    [0., 0., 0., 0.]], dtype='float64')
+    assert_array_equal(enc.fit_transform(X).toarray(), exp)
```
add here a check for the `drop_idx_` ?

jorisvandenbossche reviewed (commented):
Looking really good, thanks a lot!

Added a bunch of (minor) comments.

drewmjohnston commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -158,6 +160,21 @@ class OneHotEncoder(_BaseEncoder):
+        features cause problems, such as when feeding the resulting data
+        into a neural network or an unregularized regression. If only one
+        feature appears in a column, that column will be dropped from the
+        transformed version (unless this is overriden by manually
```
I see what you're saying--rewrote this so this only appears in the bullet point.

drewmjohnston commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -257,6 +279,12 @@ class OneHotEncoder(_BaseEncoder):
+    >>> drop_enc = OneHotEncoder(drop='first').fit(X)
+    >>> drop_enc.categories_
+    [array(['Female', 'Male'], dtype=object), array([1, 2, 3], dtype=object)]
+    >>> drop_enc.transform([['Female', 1], ['Male', 2]]).toarray()
```
As it stands, the samples used in the first example are not compatible with the drop method. The first example transforms an unknown feature (and has handle_unknown='ignore' set), but handle_unknown='ignore' is incompatible with drop, as it makes inverse transformations impossible.

drewmjohnston commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -273,7 +301,7 @@ class OneHotEncoder(_BaseEncoder):
     """
 
-    def __init__(self, n_values=None, categorical_features=None,
+    def __init__(self, n_values=None, drop=None, categorical_features=None,
```
sure thing

Commit:
incorporated changes proposed by jorisvandenbossche

drewmjohnston commented:
Alright, all of this (save the change in the examples) is incorporated now!

Commit:
Fixed flake8 error

jorisvandenbossche commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -158,6 +160,19 @@ class OneHotEncoder(_BaseEncoder):
+        features cause problems, such as when feeding the resulting data
+        into a neural network or an unregularized regression.
+
+        - None : retain all features.
```
```suggestion
        - None : retain all features (the default).
```

jorisvandenbossche commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -158,6 +160,19 @@ class OneHotEncoder(_BaseEncoder):
+
+        - None : retain all features.
+        - 'first' : drop the first category in each feature. If only one
+                    category is present, the feature will be dropped
```
rst/sphinx syntax specificity: when having a list item as this, the full text needs to be aligned with the first word, so something like:

```
- 'first' : drop the first category in each feature. If only one
  category is present, the feature will be dropped ...
```

Otherwise it will be (wrongly) formatted as a definition list, see eg https://46847-843222-gh.circle-artifacts.com/0/doc/modules/generated/sklearn.preprocessing.OneHotEncoder.html

jorisvandenbossche commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -158,6 +160,19 @@ class OneHotEncoder(_BaseEncoder):
+        - 'first' : drop the first category in each feature. If only one
+                    category is present, the feature will be dropped
+                    entirely.
+        - array : ``drop[i]`` is the caetgory in ``X[:,i]`` that should be
```
```suggestion
        - array : ``drop[i]`` is the category in feature ``X[:,i]`` that should be
```

NicolasHug replied:
nitpick: ``X[:, i]`` (with a blank space)

jorisvandenbossche commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -158,6 +160,19 @@ class OneHotEncoder(_BaseEncoder):
+                    category is present, the feature will be dropped
+                    entirely.
+        - array : ``drop[i]`` is the caetgory in ``X[:,i]`` that should be
+                  dropped. None means all categories should be retained.
```
maybe add here something like "for that feature", to make it clear this can be specified per feature (and that it is different from the None option above)

jorisvandenbossche commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -205,7 +220,13 @@ class OneHotEncoder(_BaseEncoder):
+        (if any).
+
+    drop_idx_ : array of shape (n_features,)
+        drop_idx_[i] is the index in ``categories_`` of the category to be
```
```suggestion
        drop_idx_[i] is the index in ``categories_[i]`` of the category to be
```

jorisvandenbossche commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -354,20 +383,39 @@ def _handle_deprecations(self, X):
+                        )
+                        warnings.warn(msg, FutureWarning)
+                        self._legacy_mode = True
+                        self._n_values = 'auto'
```
Can you answer this one?

NicolasHug replied:
ping @drewmjohnston

jorisvandenbossche replied:
Can you remove this line? (it is already set above)

jorisvandenbossche replied:
@drewmjohnston Can you remove this line? (or, explain why it was needed to add it?)

jorisvandenbossche commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -418,15 +472,64 @@ def fit(self, X, y=None):
+        if self.drop is not None and self.handle_unknown != 'error':
+            raise ValueError(
+                "`handle_unknown` cannot be 'ignore' when the drop parameter "
+                "is specified, as this will create ambiguous cells")
```
given that you needed to add it in two places, maybe it is worth making a helper method of it like `self._validate_keywords()` (that can also include the validation of `self.handle_unknown` that is now a few lines above, as that is also repeated)

jorisvandenbossche commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -651,18 +783,31 @@ def inverse_transform(self, X):
+                    # in each column are the dropped value
+                    if self.drop is not None:
+                        X_tr[unknown, i] = (self.categories_[i][
+                                                self.drop_idx_[i]])
```
if you re-use the already defined `cats` (instead of `self.categories_[i]`) it might fit on a single line

jorisvandenbossche commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -687,3 +705,101 @@ def test_one_hot_encoder_warning():
+                    [0., 1., 0., 1., 0.],
+                    [0., 0., 0., 0., 1.]], dtype='float64')
+    assert_array_equal(enc.fit_transform(X).toarray(), exp)
+
```
can you add here the check of `dropped_cats` that you added in the test just above (after my previous review) here as well? (then we explicitly test the None in there)

jorisvandenbossche reviewed (commented):
@drewmjohnston Thanks for the update! Added a second round of minor comments

jorisvandenbossche commented:
@NicolasHug can you give this a look as well? (since you implemented the pre-cursor version)

jorisvandenbossche commented:
@drewmjohnston can you also add an entry in the whatsnew file for 0.21 (v0.21.rst) ?

NicolasHug commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -2,11 +2,13 @@
 #          Joris Van den Bossche <jorisvandenbossche@gmail.com>
 # License: BSD 3 clause
 
+from __future__ import division
```
You shouldn't need this anymore

NicolasHug commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -2,11 +2,13 @@
 import warnings
 
 import numpy as np
+import six
```
nor this

NicolasHug reviewed (commented):
2 old imports and 1 nitpick for now, I'll try to review it more in details tomorrow.

It'd be nice to have a small text about this parameter in the User Guide, something along these lines maybe [(here)](https://github.com/scikit-learn/scikit-learn/pull/12884/files#diff-dc9a8c29395bbf58d2d77fdfddc90bb8R542)

drewmjohnston commented:
Alright, I've implemented most of these suggestions (pending further feedback on a few of them). I've added to the whatsnew and the user guide as well. Let me know what your'd like for the ones I'm unsure about.

Commit:
Added documentation for new features. Implemented revisions to tests. Fixed some nitpicks. Refactored input checking for OneHotEncoder paramenters

Commit:
Merge branch 'master' into leave_one_out_ohe

Commit:
Fixed flake8

Commit:
Fixed error check order to match expected order in pytest

NicolasHug commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -651,18 +772,31 @@ def inverse_transform(self, X):
+            else:
+                cats = np.delete(self.categories_[i], self.drop_idx_[i])
+            n_categories = len(cats)
+            if n_categories == 0:
```
Please don't remove this comment, unless it's out of date:

```

                    # Only happens if there was a column with a unique
                    # category. In this case we just fill the column with this
                    # unique category value.
```

NicolasHug commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -425,8 +475,60 @@ def fit(self, X, y=None):
+        # in interpreting the model.
+        if self.drop is not None and self.handle_unknown != 'error':
+            raise ValueError(
+                "`handle_unknown` cannot be 'ignore' when the drop parameter "
```
I'd be even more explicit and use:

> `handle_unknown` must be 'error' ...

NicolasHug commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -651,18 +772,31 @@ def inverse_transform(self, X):
-            if self.handle_unknown == 'ignore':
-                # ignored unknown categories: we have a row of all zero's
+            X_tr[:, i] = cats[labels]
+            if self.handle_unknown == 'ignore' or self.drop is not None:
```
Please consider keeping these 2 cases separate, like they are in https://github.com/scikit-learn/scikit-learn/pull/12884

These 2 cases are supposed to be mutually exclusive, and the only portion of code they have in common is the computation of 
`                unknown = np.asarray(sub.sum(axis=1) == 0).flatten()`

which is not a reason for a refactoring IMO. It's unclear at first sight why they are treated at the same place. Moreover `unknown` is not an appropriate name here if we're in the `drop is not None` case.

NicolasHug reviewed (commented):
Looks pretty good overall!

Commit:
Refactored some code, small changes from @nicolashug

Commit:
Added code example

Commit:
updated relative reference in rst to OneHotEncoder

NicolasHug commented on doc/whats_new/v0.21.rst:
```diff
@@ -273,6 +273,11 @@ Support for Python 3.4 and below has been officially dropped.
   :user:`Raffaello Baluyot <baluyotraf>`
 
+- |Feature| :class:`OneHotEncoder` now supports dropping one feature per category
+  with a new drop parameter. :issue:`6488` and :issue:`6053` by
```
The issue number should be the PR number

```suggestion
  with a new drop parameter. :issue:`12908` by
```

Also please use `:user:`

NicolasHug commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -526,8 +526,8 @@ def _validate_keywords(self):
-                "`handle_unknown` cannot be 'ignore' when the drop parameter "
-                "is specified, as this will create ambiguous cells")
+                "`handle_unknown` must be 'error' when the drop parameter "
+                "is specified to avoid ambiguities in inverse_transform")
```
It's not just about `inverse_transform`. `transform` would map both unkown and dropped categories to all zeros so the transformed output would be ambiguous.
May I suggest

```
`handle_unknown` must be 'error' when the drop parameter is
specified, as both would create categories that are all-zeros.
```

jorisvandenbossche commented on doc/modules/preprocessing.rst:
```diff
@@ -533,13 +533,29 @@ columns for this feature will be all zeros
 
 
+It is also possible to encode each column into ``n_categories - 1`` columns
+instead of ``n_categories`` columns by using the ``'drop'`` parameter. This
```
```suggestion
instead of ``n_categories`` columns by using the ``drop`` parameter. This
```

jorisvandenbossche commented on doc/modules/preprocessing.rst:
```diff
@@ -533,13 +533,29 @@ columns for this feature will be all zeros
+This is useful to avoid co-linearity in the input matrix in non-regularized
+regression (:class:`LinearRegression <sklearn.linear_model.LinearRegression>`),
+which would cause the covariance matrix to be non-invertible. In this case
+``'handle_unknown'`` must be set to ``'error'``::
```
```suggestion
``handle_unknown`` must be set to ``'error'``::
```

NicolasHug reviewed (approved):
Last comments, LGTM otherwise.

Nice work @drewmjohnston !

jorisvandenbossche commented on doc/modules/preprocessing.rst:
```diff
@@ -533,13 +533,29 @@ columns for this feature will be all zeros
+instead of ``n_categories`` columns by using the ``'drop'`` parameter. This
+parameter allows the user to specify a category for each feature to be dropped.
+This is useful to avoid co-linearity in the input matrix in non-regularized
+regression (:class:`LinearRegression <sklearn.linear_model.LinearRegression>`),
```
maybe state this a bit broader? (or explicitly make it clear the LinearRegression class is an example, as now it seems that this is the only use case)

NicolasHug commented:
Closes #12884

jorisvandenbossche commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -205,7 +218,13 @@ class OneHotEncoder(_BaseEncoder):
+        (if any).
+
+    drop_idx_ : array of shape (n_features,)
+        drop_idx_[i] is the index in ``categories_[i]`` of the category to be
```
```suggestion
        ``drop_idx_[i]`` is the index in ``categories_[i]`` of the category to be
```

jorisvandenbossche commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -651,18 +772,36 @@ def inverse_transform(self, X):
                 unknown = np.asarray(sub.sum(axis=1) == 0).flatten()
                 if unknown.any():
+                    # ignored unknown categories: we have a row of all
+                    # zero
```
can you lave this comment line where it was before? (then it will also fit on a single line again)

jorisvandenbossche reviewed (commented):
few more minor details (but again, looking very good, should be my last round I think)

jorisvandenbossche commented:
> Alright, I've implemented most of these suggestions (pending further feedback on a few of them).

BTW, make sure to ask for clarification by answering to a review comment if you need further feedback on one of them

jorisvandenbossche commented:
One API related aspect I was wondering: we currently enable people to specify feature-by-feature (the case where `drop` is an array) to drop a specific category, or to drop no categories. But, we don't enable users to choose for 'first' for a specific feature. 
This is of course difficult to enable since 'first' could conflict with an actual category, but this gives a slight inconsistency compared to None (both 'first' and None are scalar options for `drop`, but only None can be included in the array).

Probably not the most important remark, just wanted to note it.

drewmjohnston commented:
@jorisvandenbossche I agree that that functionality could perhaps be valuable to some users. However, given that it'd be tricky to implement (given the possibility that 'first' is a category), I think it might be better to let users implement it themselves. If a user's specifying 'first' for a given feature, they probably don't care about the interpretation of the coefficients on the different categories. To them, dropping the first value of a feature (by sorting) or the first value of the feature in the array (which they could grab with a list comprehension) is probably interchangeable.

Commit:
Changes to documentation for OHE, changes to legacy implementation to reflect that it's being merged into 0.22

Commit:
Merge branch 'master' into leave_one_out_ohe

jnothman commented:
Discussing with @jorisvandenbossche, we think it's unnecessary complexity to support None being in the list of drops.... this can be achieved, if the user needs it, with a ColumnTransformer. Supporting None in the list adds unnecessary complexity to the code, and may make it hard to support having None be a valid (or missing value) category in the input in the future. I hope it's not too much effort, @drewmjohnston, to pull back that feature.

drewmjohnston commented:
That's not a problem--it'll make the code cleaner and allow us to check the dtype of drop_idx too. I'll incorporate this and re-push.

Commit:
Removed option to select drop-one only on some columns. This will be implemented with FeatureTransformer

Commit:
Merge branch 'master' into leave_one_out_ohe

Commit:
Reformatting, fixed flake8

jnothman reviewed (approved):
Thanks for making those changes, @drewmjohnston

Commit:
Small doc fix

jorisvandenbossche reviewed (commented):
Latest changes look all good, thanks!

jorisvandenbossche commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -571,11 +651,23 @@ def _transform_new(self, X):
         X_int, X_mask = self._transform(X, handle_unknown=self.handle_unknown)
 
+        if self.drop is not None:
+            max_val = max(len(cats) for cats in self.categories_)
```
Ignore this one, wasn't looking at the diff with the last commit

drewmjohnston commented:
Great--made the small documentation fix you recommended.

jorisvandenbossche commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -372,47 +372,30 @@ def _handle_deprecations(self, X):
-                            "an array of entirely integer data. This will "
-                            "enable the future behavior."
-                        )
-                        raise ValueError(msg)
```
@drewmjohnston what is the reason you removed this ValueError again (for the case where we would normally end up in the legacy behaviour, but where the user did specify the `drop` keyword).

jorisvandenbossche replied:
Just tried this out locally. So you still get an error, but the one related to `n_values` or `categorical_features` not being compatible with `drop`. Which is a bit confusing here, since you didn't specify those. So I would either add it back, or change the other error message a bit.

Commit:
Fixed n_values/drop compatibility

drewmjohnston commented:
@jorisvandenbossche OK, sounds good. I removed this in a moment of confusion when I though this was going in to .22, not .21.

jorisvandenbossche commented:
@drewmjohnston Can you also add back the test? (it's in here: https://github.com/scikit-learn/scikit-learn/pull/12908/commits/2eb2c1699ba431adfc8ee08b6a77199b635127f1)

Commit:
Removed cruft. Added in test for drop/n_values interaction

jorisvandenbossche commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -346,28 +370,46 @@ def _handle_deprecations(self, X):
+                            "[0, max(values)], while in the future they "
+                            "will be determined based on the unique "
+                            "values.\n The old behavior is not compatible "
+                            "with the `drop` paramenter. Instead, you "
```
```suggestion
                            "with the `drop` parameter. Instead, you "
```

drewmjohnston commented:
OK, removed the extra code and added the tests for the drop/n_values interactions.

jorisvandenbossche reviewed (approved):
OK, looks all good now!

amueller commented:
merging as it has two +1s and appveyor is kinda redundant now.

NicolasHug commented:
You didn't merge @amueller ?

jorisvandenbossche commented:
@drewmjohnston Thanks a lot!

drewmjohnston commented:
Great! Thanks for the help everyone. You guess make contributing to this a breeze.
